In [1]:
import pandas as pd
import numpy as np
import scipy.stats as st
import statsmodels.api as sm

import seaborn as sns
import matplotlib.pyplot as plt

pd.set_option('max_colwidth', 200)
pd.set_option('display.width', 400)

from IPython.core.display import display, HTML
display(HTML("<style>.container { width:70% !important; }</style>"))

# Тест на оценку

In [7]:
df = pd.read_csv('../Data/CandyData.csv')
df.head(3)

competitorname  chocolate  fruity  caramel  peanutyalmondy  nougat  crispedricewafer  hard  bar  pluribus  sugarpercent  pricepercent  winpercent
0      100 Grand          1       0        1               0       0                 1     0    1         0         0.732         0.860   66.971725
1   3 Musketeers          1       0        0               0       1                 0     0    1         0         0.604         0.511   67.602936
2       One dime          0       0        0               0       0                 0     0    0         0         0.011         0.116   32.261086

## Описание данных

- chocolate/fruity/caramel/peanutalmondy/nougat/crispedricewafer — содержит ли конфета шоколад/фруктовый ароматизатор/карамель/арахис/нугу/воздушный рис (1 – да/ 0 – нет)
- hard — конфета твердая (1 – да/ 0 – нет)
- bar — конфета похожа на батончик (1 – да/ 0 – нет)
- pluribus — в упаковке одна конфета или несколько (1 – несколько/ 0 – одна)
- sugarpercent — процент сахара в конфете
- pricepercent — цена конфеты
- winpercent — рейтинг конфеты

In [8]:
df['winpercent_0_1'] = np.where(df['winpercent'] >= 50, 1, 0)

In [9]:
df.head()

competitorname  chocolate  fruity  caramel  peanutyalmondy  nougat  crispedricewafer  hard  bar  pluribus  sugarpercent  pricepercent  winpercent  winpercent_0_1
0      100 Grand          1       0        1               0       0                 1     0    1         0         0.732         0.860   66.971725               1
1   3 Musketeers          1       0        0               0       1                 0     0    1         0         0.604         0.511   67.602936               1
2       One dime          0       0        0               0       0                 0     0    0         0         0.011         0.116   32.261086               0
3    One quarter          0       0        0               0       0                 0     0    0         0         0.011         0.511   46.116505               0
4      Air Heads          0       1        0               0       0                 0     0    0         0         0.906         0.511   52.341465               1

## Задания

### Постройте логистическую модель, где зависимой переменной выступает `winpercent_0_1`, а независимой – `sugarpercent`. Выберите верную интерпретацию константы.

In [13]:
X = df[['sugarpercent']].assign(const=1)
y = df['winpercent_0_1']
model_logit = sm.Logit(y, X)
result = model_logit.fit()
print(result.summary())

Optimization terminated successfully.
         Current function value: 0.672972
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:         winpercent_0_1   No. Observations:                   85
Model:                          Logit   Df Residuals:                       83
Method:                           MLE   Df Model:                            1
Date:                Sun, 05 Jun 2022   Pseudo R-squ.:                 0.02433
Time:                        11:01:52   Log-Likelihood:                -57.203
converged:                       True   LL-Null:                       -58.629
Covariance Type:            nonrobust   LLR p-value:                   0.09122
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
sugarpercent     1.3276      0.798      1.663      0.096      -0.237       2.892
const           -0.8058

In [17]:
np.exp(-0.8058)

0.4467303992481016

 - [ ] При равным нулю процентном содержании сахара шанс, что у конфеты будет высокий рейтинг, ниже в 2.23 раза
 - [x] При равным нулю процентном содержании сахара шанс, что у конфеты будет высокий рейтинг, ниже в 0.44 раза
 - [ ] При равным нулю процентном содержании сахара шанс, что у конфеты будет низкий рейтинг, ниже в 2.23 раза
 - [ ] При равным нулю процентном содержании сахара шанс, что у конфеты будет низкий рейтинг, ниже в 0.44 раза

### Постройте логистическую модель, где зависимой переменной выступает `winpercent_0_1`, а независимой – `sugarpercent`. Выберите верную интерпретацию коэффициента при объясняющей переменной. Выберите один верный ответ.

In [18]:
np.exp(1.3276)

3.771979764137272

- [x] При увеличении процента содержания сахара на 1 шанс, что у конфеты будет высокий рейтинг, увеличивается в 3.77 раза
- [ ] При увеличении процента содержания сахара на 1 шанс, что у конфеты будет высокий рейтинг, увеличивается в 1.32 раза
- [ ] При увеличении процента содержания сахара на 1 шанс, что у конфеты будет низкий рейтинг, увеличивается в 3.77 раза
- [ ] При увеличении процента содержания сахара на 1 шанс, что у конфеты будет низкий рейтинг, увеличивается в 1.32 раза

### Постройте логистическую модель, где зависимой переменной выступает `winpercent_0_1`, а независимой – `sugarpercent`. Укажите, верно ли следующее утверждение: на уровне значимости 0.05 процент сахара в конфете статистически значимо влияет на рейтинг конфеты.

### Постройте логистическую модель, где зависимой переменной выступает `winpercent_0_1`, а независимой – `sugarpercent`. Чему будет равен псевдо R-квадрат в данной модели? Выберите один верный ответ.

In [19]:
0.02433

0.02433

### Постройте логистическую модель, где зависимой переменной выступает `winpercent_0_1`, а независимыми – `sugarpercent` и `pricepercent`. Чему будет равено отношение шансов при цене конфеты? Введите ответ в виде десятичной дроби, округлите до двух знаков после запятой.

In [23]:
X = df[['sugarpercent', 'pricepercent']].assign(const=1)
y = df['winpercent_0_1']
model = sm.Logit(y, X)
result = model.fit()
print(result.summary())

Optimization terminated successfully.
         Current function value: 0.619123
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:         winpercent_0_1   No. Observations:                   85
Model:                          Logit   Df Residuals:                       82
Method:                           MLE   Df Model:                            2
Date:                Sun, 05 Jun 2022   Pseudo R-squ.:                  0.1024
Time:                        11:16:05   Log-Likelihood:                -52.625
converged:                       True   LL-Null:                       -58.629
Covariance Type:            nonrobust   LLR p-value:                  0.002470
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
sugarpercent     0.5896      0.881      0.669      0.503      -1.137       2.316
pricepercent     2.6012

In [25]:
oods_sugar = np.exp(0.5896)
oods_price= np.exp(2.6012)

result = oods_sugar / oods_price
result

0.13377446420340544

### Постройте логистическую модель, где зависимой переменной выступает `winpercent_0_1`, а независимыми – `sugarpercent` и `pricepercent`. Чему будет равно отношение шансов при проценте сахара в конфете? Введите ответ в виде десятичной дроби, округлите до двух знаков после запятой.

In [28]:
result = oods_price / oods_sugar
result.round(2)

7.48